pdns/pdnsutil.cc

```diff
@@ -1673,6 +1673,86 @@ bool showZone(DNSSECKeeper& dk, const DNSName& zone)
   return true;
 }
 
+bool exportZoneDS(DNSSECKeeper& dk, const DNSName& zone)
+{
+  UeberBackend B("default");
+  DomainInfo di;
+
+  static const std::vector<std::pair<unsigned int, std::string> > DS_HASH_NAMES = {
+    std::make_pair(1, "SHA1"),
+    std::make_pair(2, "SHA256"),
+    std::make_pair(3, "GOST R 34.11-94"),
+    std::make_pair(4, "SHA384"),
+  };
+
+  if (!B.getDomainInfo(zone, di)) {
+    cerr << "No such zone in the database" << endl;
+    return false;
+  }
+
+  if (!dk.isSecuredZone(zone)) {
+    cerr << "Zone is not actively secured" << endl;
+    return false;
+  }
+
+  vector<DNSKEYRecordContent> keys;
+
+  if (dk.isPresigned(zone)) {
+    // get us some keys
+    vector<DNSKEYRecordContent> keys;
+    DNSResourceRecord rr;
+
+    B.lookup(QType(QType::DNSKEY), zone);
+    while(B.get(rr)) {
+      if (rr.qtype != QType::DNSKEY) continue;
+      keys.push_back(*dynamic_cast<DNSKEYRecordContent*>(DNSKEYRecordContent::make(rr.getZoneRepresentation())));
+    }
+
+  } else {
+    DNSSECKeeper::keyset_t keyset = dk.getKeys(zone);
+    for (const auto &value: keyset) {
+      keys.push_back(value.first.getDNSKEY());
+    }
+
+  }
+
+  if(keys.empty()) {
+    cerr << "No keys for zone '" << zone.toString() << "'."<<endl;
+    return true;
+  }
+
+  sort(keys.begin(),keys.end());
+  reverse(keys.begin(),keys.end());
+  for(const auto& key : keys) {
+    if (key.d_flags != 257) {
```

Comment:
On zones with ONLY flags=256 keys, you should print those. I'm pretty sure you can reuse this logic from somewhere else.